cob_cartesian_controller/src/cartesian_controller.cpp

```diff
@@ -29,430 +29,366 @@
 {
-    tf::TransformBroadcaster br;
-    tf::Transform transform;
-    tf::TransformListener listener;
-    reached_pos_ = false;
+    bool success = false;
     int reached_pos_counter = 0;
-    double ro, pi, ya;
+
     ros::Rate rate(update_rate_);
     tf::StampedTransform stamped_transform;
-    tf::Quaternion q;
-
-    while(ros::ok())
+    stamped_transform.setOrigin( tf::Vector3(target_pose.position.x,
+                                             target_pose.position.y,
+                                             target_pose.position.z) );
+    stamped_transform.setRotation( tf::Quaternion(target_pose.orientation.x,
+                                                  target_pose.orientation.y,
+                                                  target_pose.orientation.z,
+                                                  target_pose.orientation.w) );
+    stamped_transform.frame_id_ = root_frame_;
+    stamped_transform.child_frame_id_ = target_frame_;
+
+    while(as_->isActive())
     {
-        // Linearkoordinaten
-        transform.setOrigin( tf::Vector3(target_pose.position.x, target_pose.position.y, target_pose.position.z) );
-
-        q = tf::Quaternion(target_pose.orientation.x, target_pose.orientation.y, target_pose.orientation.z, target_pose.orientation.w);
-        transform.setRotation(q);
-
-        // Send br Frame
-        br.sendTransform(tf::StampedTransform(transform, ros::Time::now(), reference_frame_, target_frame_));
+        // Send/Refresh target Frame
+        stamped_transform.stamp_ = ros::Time::now();
+        tf_broadcaster_.sendTransform(stamped_transform);
 
         // Get transformation
-        try
-        {
-            startTracking();
-            listener.lookupTransform(target_frame_, chain_tip_link_, ros::Time(0), stamped_transform);
-        }
-        catch (tf::TransformException &ex)
-        {
-            ROS_ERROR("%s",ex.what());
-        }
-
-        // Get current RPY out of quaternion
-        tf::Quaternion quatern = stamped_transform.getRotation();
-        tf::Matrix3x3(quatern).getRPY(ro,pi,ya);
-
+        stamped_transform = utils_.getStampedTransform(target_frame_, chain_tip_link_);
 
-        // Wait for arm_7_link to be in position
-        if(utils_.epsilonArea(stamped_transform.getOrigin().x(), stamped_transform.getOrigin().y(), stamped_transform.getOrigin().z(), ro, pi, ya,epsilon))
+        // Wait for chain_tip_link to be within epsilon area of target_frame
+        if(utils_.inEpsilonArea(stamped_transform, epsilon))
         {
             reached_pos_counter++;    // Count up if end effector position is in the epsilon area to avoid wrong values
         }
 
-        if(reached_pos_counter >= 50)
-        {
-            reached_pos_ = true;
-        }
-
-        if(reached_pos_ == true)    // Cancel while loop
+        if(reached_pos_counter >= 2*update_rate_) //has been close enough to target for 2 seconds
```

Comment:
These kind of "reached"-check (and also the inEpsilonArea-check) should be handled by the cob_frame_tracker ActionServer....the Server should then decide when the target has been reached and report success (or failure in case the target was not reaches - after a timeout - or anything else....). I would like to get rid of these kind of checks from the cob_cartesian_controller in the end...